### PR TITLE
LH-69472: Require that the API token provided is for a user with ADMIN or SUPER_ADMIN privileges

### PR DESCRIPTION
**Pull request by hazital**
https://jira-eng-rtp3.cisco.com/jira/browse/LH-69472

### Description
Add a validator for the provider to ensure that only users with the roles of ADMIN or SUPER_ADMIN are able to perform actions

### CODE REVIEW

File: provider/examples/resources/ios/ios_example.tf [hunk 1]
```diff
@@ -0,0 +1,21 @@
+terraform {
+  required_providers {
+    cdo = {
+      source = "hashicorp.com/CiscoDevnet/cdo"
+    }
+  }
+}
+
+provider "cdo" {
+  base_url  = "https://staging.dev.lockhart.io"
```
Comment (siddhuwarrier) [inline provider/examples/resources/ios/ios_example.tf:10]: Hey please take a look at asa_example.tf (unless my changes were overwritten), I wrote slightly better placeholder text than `FILL_ME`. Could you replicate that please?

File: provider/internal/device/asa/data_source.go [hunk 2]
```diff
@@ -74,7 +74,7 @@ func (d *AsaDataSource) Schema(ctx context.Context, req datasource.SchemaRequest
 				MarkdownDescription: "The type of the connector that is used to communicate with the device. CDO can communicate with your device using either a Cloud Connector (CDG) or a Secure Device Connector (SDC); see [the CDO documentation](https://docs.defenseorchestrator.com/c-connect-cisco-defense-orchestratortor-the-secure-device-connector.html) to learn mor (Valid values: [CDG, SDC]).",
 				Computed:            true,
 				Validators: []validator.String{
-					validators.OneOf("CDG", "SDC"),
+					stringvalidator.OneOf("CDG", "SDC"),
```
Comment (siddhuwarrier) [inline provider/internal/device/asa/data_source.go:77]: Nice find!

File: provider/validators/cdo_role.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+type oneOfRolesValidator struct {
+	expectedRoles []types.String
+}
+
+func (v oneOfRolesValidator) Description(ctx context.Context) string {
+	return v.MarkdownDescription(ctx)
+}
+
+func (v oneOfRolesValidator) MarkdownDescription(_ context.Context) string {
+	return fmt.Sprintf("must contains one of CDO roles: %q", v.expectedRoles)
```
Comment (siddhuwarrier) [inline provider/validators/cdo_role.go:34]: Reword to "The user must be assigned one of the following CDO roles: "?

File: provider/validators/cdo_role.go [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+	if request.ConfigValue.IsNull() || request.ConfigValue.IsUnknown() {
+		return
+	}
+
+	token := request.ConfigValue.String()
+
+	role, err := extractRoleFromToken(token)
+
+	if err != nil {
+		fmt.Println("Error:", err)
```
Comment (siddhuwarrier) [inline provider/validators/cdo_role.go:47]: The indentation looks a bit off here.

File: provider/validators/cdo_role.go [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+	}
+
+	for _, expectedRole := range v.expectedRoles {
+		if role == expectedRole.ValueString() {
+			return
```
Comment (siddhuwarrier) [inline provider/validators/cdo_role.go:52]: Same. Maybe it's just Github?

File: provider/validators/cdo_role.go [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+	}
+
+	response.Diagnostics.Append(validatordiag.InvalidAttributeValueMatchDiagnostic(
+		request.Path,
+		v.Description(ctx),
+		role,
+	))
+}
+
+func extractRoleFromToken(tokenString string) (string, error) {
```
Comment (siddhuwarrier) [inline provider/validators/cdo_role.go:63]: Shouldn't we do a nil check here? Or will that be caught earlier?

File: provider/validators/cdo_role.go [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+		return "", fmt.Errorf(ErrDecodeFailed)
+	}
+
+	var claims jwt.MapClaims
+	if err := json.Unmarshal(payloadBytes, &claims); err != nil {
+		return "", fmt.Errorf("failed to decode token payload: %v", err)
+	}
+
+	if rolesClaim, exists := claims["roles"]; exists {
+		if roles, ok := rolesClaim.([]interface{}); ok && len(roles) > 0 {
```
Comment (siddhuwarrier) [inline provider/validators/cdo_role.go:80]: This is quite hard to parse. Could we please split this into multiple statements so that it's easier to read?

File: provider/validators/cdo_role.go [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+			if role, ok := roles[0].(string); ok {
```
Comment (siddhuwarrier) [inline provider/validators/cdo_role.go:81]: Same comment. Also, why are we only getting the first role?

File: provider/validators/cdo_role_test.go [hunk 9]
```diff
@@ -1,72 +1,56 @@
...
 	"github.com/hashicorp/terraform-plugin-framework/types"
-
-	"github.com/hashicorp/terraform-plugin-framework-validators/stringvalidator"
 )
 
-func TestOneOfValidator(t *testing.T) {
+func TestOneOfRolesValidator(t *testing.T) {
 	t.Parallel()
 
+	SuperAdminRoleToken := "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJyb2xlcyI6WyJST0xFX1NVUEVSX0FETUlOIl19.FBEPzCpXPiYX6esud26WMrJvU3reLDVFdLociGrilVw"
```
Comment (siddhuwarrier) [inline provider/validators/cdo_role_test.go:15]: Just checking: these tokens are now expired, right? If not (i.e., if they are API tokens), please revoke them and use temporary access tokens that we know will expire.

**Review by siddhuwarrier — CHANGES_REQUESTED** [review]
I really like the overall approach; just a few code comments.

File: provider/validators/cdo_role.go [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+
+	response.Diagnostics.Append(validatordiag.InvalidAttributeValueMatchDiagnostic(
+		request.Path,
+		v.Description(ctx),
+		role,
+	))
+}
+
+func extractRoleFromToken(tokenString string) (string, error) {
+	parts := strings.Split(tokenString, ".")
```
Comment (weilueluo) [inline provider/validators/cdo_role.go:64]: maybe use `jwt.Parse` instead? https://pkg.go.dev/github.com/golang-jwt/jwt/v5#example-Parse-Hmac

**Review by weilueluo — CHANGES_REQUESTED** [review]
Looks good! Nice, just one comment

File: provider/examples/resources/ios/ios_example.tf (hunk 1 above)
Comment (hazital) [reply provider/examples/resources/ios/ios_example.tf:10]: sure thanks

File: provider/validators/cdo_role.go (hunk 4 above)
Comment (hazital) [reply provider/validators/cdo_role.go:47]: Actually there's only one tab, not sure why it appears that way on Github

File: provider/validators/cdo_role.go (hunk 5 above)
Comment (hazital) [reply provider/validators/cdo_role.go:52]: yeah It's just Github

File: provider/validators/cdo_role_test.go (hunk 9 above)
Comment (hazital) [reply provider/validators/cdo_role_test.go:15]: Yes, these are fake JWT tokens that only contain the data I need for the tests

File: provider/validators/cdo_role.go (hunk 6 above)
Comment (hazital) [reply provider/validators/cdo_role.go:63]: sure will add, thanks

File: provider/validators/cdo_role.go (hunk 8 above)
Comment (hazital) [reply provider/validators/cdo_role.go:81]: Because the user can have only one role, is it not true?

File: provider/validators/cdo_role.go (hunk 10 above)
Comment (hazital) [reply provider/validators/cdo_role.go:64]: I didn't use the `jwt.Parse` since it requires providing the secret key and I preferred to keep it simple using base64 encoding

File: provider/examples/resources/ios/ios_example.tf (hunk 1 above)
Comment (weilueluo) [inline provider/examples/resources/ios/ios_example.tf:10]: Do we want to change this?

File: provider/validators/cdo_role.go (hunk 8 above)
Comment (weilueluo) [reply provider/validators/cdo_role.go]: Let's we be more defensive --- as long as one of the user role is `ADMIN` or `SUPERADMIN` we will allow it.

**Review by weilueluo — APPROVED** [review]
Nice work! Approved but please change verification of the role if it sounds right.

File: provider/validators/cdo_role.go (hunk 8 above)
Comment (hazital) [reply provider/validators/cdo_role.go]: @weilueluo  Actually, this method only extracts the role from the token. The validation occurs at the main provider, and I'm verifying that the role is precisely one of them using `validators.OneOfRoles("ROLE_SUPER_ADMIN", "ROLE_ADMIN"),`

Comment (weilueluo) [reply provider/validators/cdo_role.go]: Ah right, I see, it would be a pain to extract multiple role, make sense

File: provider/examples/resources/ios/ios_example.tf (hunk 1 above)
Comment (hazital) [reply provider/examples/resources/ios/ios_example.tf]: yep thanks
